File: string matching/KMP/KMP.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

// This function computes the partial match table for the given pattern
vector<int> kmpTable(string pattern) {
    int m = pattern.length();
    vector<int> pi(m);

    int k = 0;
    for (int i = 1; i < m; i++) {
        // This loop updates the partial match table based on the current character of the pattern
        // If there is a partial match, it updates the table with the length of the matching prefix
        // If there is no partial match, it backtracks to the longest prefix that has a partial match
        while (k > 0 && pattern[k] != pattern[i]) {
            k = pi[k-1];
        }
        if (pattern[k] == pattern[i]) {
            k++;
        }
        pi[i] = k;
    }
    return pi;
}

// This function searches for the given pattern in the given text using the KMP algorithm
int kmpSearch(string text, string pattern) {
    int n = text.length();
    int m = pattern.length();

    // Compute the partial match table for the pattern
    vector<int> pi = kmpTable(pattern);

    int i = 0, j = 0;
    while (i < n) {
        // This loop compares each character of the text and the pattern until a match is found or the end of the text is reached
        if (text[i] == pattern[j]) {
            i++;
            j++;
            if (j == m) {
                return i - j; // If the entire pattern is matched, return the starting position of the match
            }
        } else if (j > 0) {
            j = pi[j-1]; // If there is a partial match, backtrack to the longest prefix that has a partial match
        } else {
            i++; // If there is no partial match, move to the next character in the text
        }
    }
    return -1; // If the pattern is not found, return -1
}

int main() {
    string text = "Hello, world!";
    string pattern = "world";
    int pos = kmpSearch(text, pattern);
    if (pos == -1) {
        cout << "Pattern not found" << endl;
    } else {
        cout << "Pattern found at position " << pos << endl;
    }
    return 0;
}

File: string matching/KMP/KMP.py
def kmp_table(pattern):
    # Create a table to store the values of the longest proper prefix that is also a suffix of the substring for each position in the pattern.
    table = [0] * len(pattern)
    
    # Initialize the left and right pointers to zero and one, respectively.
    left, right = 0, 1
    
    # Iterate over the pattern from left to right.
    while right < len(pattern):
        # If the character at the right pointer is equal to the character at the left pointer, increment both pointers and set the value of the table at the right pointer to the value of the left pointer.
        if pattern[right] == pattern[left]:
            left += 1
            table[right] = left
            right += 1
        else:
            # If the characters are not equal, move the left pointer back to the position in the table corresponding to the previous longest proper prefix that is also a suffix, and continue checking for a match.
            if left != 0:
                left = table[left-1]
            else:
                # If there is no previous longest proper prefix that is also a suffix, set the value of the table at the right pointer to zero and move the right pointer forward.
                table[right] = 0
                right += 1
    
    return table


def kmp_search(text, pattern):
    # Create a table to store the values of the longest proper prefix that is also a suffix of the substring for each position in the pattern.
    table = kmp_table(pattern)
    
    # Initialize variables for the indices of the text and pattern.
    i, j = 0, 0
    
    # Iterate over the text while the index is less than the length of the text.
    while i < len(text):
        # If the characters at the current indices match, increment both indices.
        if text[i] == pattern[j]:
            i += 1
            j += 1
            
            # If the value of j is equal to the length of the pattern, the pattern has been found in the text, so return the index where it starts.
            if j == len(pattern):
                return i - j
        else:
            # If the characters do not match and j is not zero, move the j index to the value in the table corresponding to the previous longest proper prefix that is also a suffix, and continue checking for a match.
            if j != 0:
                j = table[j-1]
            else:
                # If there is no previous longest proper prefix that is also a suffix, move the i index forward.
                i += 1
    
    # If the pattern is not found, return -1.
    return -1

text = "Hello, world"
pattern = "world"

pos = kmp_search(text, pattern);

if pos == -1:
    print("Pattern not found")
else:
    print("Pattern found at position", pos)